File: cli/src/dfu.rs
use rusb::{Context, DeviceHandle, Direction, Recipient, RequestType, UsbContext};
use std::{thread, time::Duration};

pub const DEFAULT_USB_VENDOR_ID: u16 = 0x0483;
pub const DEFAULT_USB_PRODUCT_ID: u16 = 0xDF11;

const STATUS_OK: u8 = 0x00;

const DFU_DNLOAD: u8 = 0x01;
const DFU_UPLOAD: u8 = 0x02;
const DFU_GETSTATUS: u8 = 0x03;
const DFU_CLRSTATUS: u8 = 0x04;

const STATE_DFU_IDLE: u8 = 0x02;
const STATE_DFU_DNLOAD_SYNC: u8 = 0x03;
const STATE_DFU_DOWNLOAD_BUSY: u8 = 0x04;
const STATE_DFU_DOWNLOAD_IDLE: u8 = 0x05;
const STATE_DFU_MANIFEST_SYNC: u8 = 0x06;
const STATE_DFU_MANIFEST: u8 = 0x07;
const STATE_DFU_MANIFEST_WAIT_RESET: u8 = 0x08;
const STATE_DFU_UPLOAD_IDLE: u8 = 0x09;
const STATE_DFU_ERROR: u8 = 0x0A;

pub const BLOCK_SIZE: usize = 2048;
const DFU_INTERFACE: u16 = 0;

pub struct DfuDevice {
    _ctx: Context,
    handle: DeviceHandle<Context>,
}

impl DfuDevice {
    pub fn open(vendor_id: u16, product_id: u16) -> Result<Self, String> {
        let ctx = Context::new().map_err(|e| format!("Failed to init USB context: {e}"))?;
        let handle = ctx
            .open_device_with_vid_pid(vendor_id, product_id)
            .ok_or_else(|| format!("No DFU device found (VID 0x{vendor_id:04x}, PID 0x{product_id:04x})"))?;

        let mut device = Self { _ctx: ctx, handle };

        let _ = device.handle.set_auto_detach_kernel_driver(true);
        let _ = device.handle.set_active_configuration(1);
        let _ = device
            .handle
            .set_alternate_setting(DFU_INTERFACE as u8, 0);

        device
            .handle
            .claim_interface(DFU_INTERFACE as u8)
            .map_err(|e| format!("Failed to claim DFU interface {DFU_INTERFACE}: {e}"))?;

        // Some DFU implementations can start in dfuERROR; clear it up-front.
        if let Ok(status) = device.get_status() {
            if status[4] == STATE_DFU_ERROR {
                let _ = device.clear_status();
            }
        }

        Ok(device)
    }

    fn req_type_out() -> u8 {
        rusb::request_type(Direction::Out, RequestType::Class, Recipient::Interface)
    }

    fn req_type_in() -> u8 {
        rusb::request_type(Direction::In, RequestType::Class, Recipient::Interface)
    }

    fn control_out(
        &mut self,
        request: u8,
        value: u16,
        data: &[u8],
        timeout_ms: u64,
    ) -> Result<usize, String> {
        self.handle
            .write_control(
                Self::req_type_out(),
                request,
                value,
                DFU_INTERFACE,
                data,
                Duration::from_millis(timeout_ms),
            )
            .map(|n| n as usize)
            .map_err(|e| format!("USB control OUT failed (req=0x{request:02x}): {e}"))
    }

    fn control_in(
        &mut self,
        request: u8,
        value: u16,
        data: &mut [u8],
        timeout_ms: u64,
    ) -> Result<usize, String> {
        self.handle
            .read_control(
                Self::req_type_in(),
                request,
                value,
                DFU_INTERFACE,
                data,
                Duration::from_millis(timeout_ms),
            )
            .map(|n| n as usize)
            .map_err(|e| format!("USB control IN failed (req=0x{request:02x}): {e}"))
    }

    pub fn get_status(&mut self) -> Result<[u8; 6], String> {
        let mut buf = [0u8; 6];
        let n = self.control_in(DFU_GETSTATUS, 0, &mut buf, 500)?;
        if n < 6 {
            return Err(format!("DFU_GETSTATUS returned {n} bytes (expected 6)"));
        }
        Ok(buf)
    }

    pub fn clear_status(&mut self) -> Result<(), String> {
        self.control_out(DFU_CLRSTATUS, 0, &[], 5000)?;
        Ok(())
    }

    pub fn wait_download_idle(&mut self) -> Result<(), String> {
        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(5);

        let mut status = self.get_status()?;
        while !(status[4] == STATE_DFU_IDLE || status[4] == STATE_DFU_DOWNLOAD_IDLE) {
            if start.elapsed() > timeout {
                return Err(format!(
                    "Timeout exceeded while waiting for download idle state (status={})",
                    format_status(status)
                ));
            }
            self.clear_status()?;
            status = self.get_status()?;
        }
        Ok(())
    }

    pub fn wait_upload_idle(&mut self) -> Result<(), String> {
        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(5);

        let mut status = self.get_status()?;
        while !(status[4] == STATE_DFU_IDLE || status[4] == STATE_DFU_UPLOAD_IDLE) {
            if start.elapsed() > timeout {
                return Err(format!(
                    "Timeout exceeded while waiting for upload idle state (status={})",
                    format_status(status)
                ));
            }
            self.clear_status()?;
            status = self.get_status()?;
        }
        Ok(())
    }

    fn poll_timeout_ms(status: [u8; 6]) -> u64 {
        let bw_poll_timeout: u32 =
            ((status[3] as u32) << 16) | ((status[2] as u32) << 8) | (status[1] as u32);
        bw_poll_timeout as u64
    }

    pub fn mass_erase(&mut self) -> Result<(), String> {
        self.wait_download_idle()?;

        let command = [0x41u8];
        self.control_out(DFU_DNLOAD, 0, &command, 50)?;

        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(60);

        loop {
            let status = self.get_status()?;
            if status[0] != STATUS_OK || status[4] == STATE_DFU_ERROR {
                return Err(format!("Mass erase failed (status={})", format_status(status)));
            }

            match status[4] {
                STATE_DFU_IDLE | STATE_DFU_DOWNLOAD_IDLE => break,
                STATE_DFU_DNLOAD_SYNC
                | STATE_DFU_DOWNLOAD_BUSY
                | STATE_DFU_MANIFEST_SYNC
                | STATE_DFU_MANIFEST
                | STATE_DFU_MANIFEST_WAIT_RESET => {
                    if start.elapsed() > timeout {
                        return Err(format!(
                            "Timeout exceeded while waiting for mass erase (status={})",
                            format_status(status)
                        ));
                    }
                    thread::sleep(Duration::from_millis(Self::poll_timeout_ms(status).max(10)));
                }
                other => {
                    return Err(format!(
                        "Mass erase failed (unexpected DFU state 0x{other:02x}, status={})",
                        format_status(status)
                    ));
                }
            }
        }

        Ok(())
    }

    pub fn set_address_pointer(&mut self, address: u32) -> Result<(), String> {
        self.wait_download_idle()?;

        let buffer = [
            0x21u8,
            (address & 0xFF) as u8,
            ((address >> 8) & 0xFF) as u8,
            ((address >> 16) & 0xFF) as u8,
            ((address >> 24) & 0xFF) as u8,
        ];

        self.control_out(DFU_DNLOAD, 0, &buffer, 50)?;

        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(5);

        loop {
            let status = self.get_status()?;
            if status[0] != STATUS_OK || status[4] == STATE_DFU_ERROR {
                return Err(format!(
                    "Set address pointer failed (status={})",
                    format_status(status)
                ));
            }
            match status[4] {
                STATE_DFU_IDLE | STATE_DFU_DOWNLOAD_IDLE => break,
                STATE_DFU_DNLOAD_SYNC | STATE_DFU_DOWNLOAD_BUSY => {
                    if start.elapsed() > timeout {
                        return Err(format!(
                            "Timeout exceeded while setting address pointer (status={})",
                            format_status(status)
                        ));
                    }
                    thread::sleep(Duration::from_millis(Self::poll_timeout_ms(status).max(10)));
                }
                other => {
                    return Err(format!(
                        "Set address pointer failed (unexpected DFU state 0x{other:02x}, status={})",
                        format_status(status)
                    ));
                }
            }
        }

        Ok(())
    }

    pub fn write_block(&mut self, block_num: u16, data: &[u8]) -> Result<(), String> {
        self.wait_download_idle()?;
        self.control_out(DFU_DNLOAD, block_num, data, 500)?;

        let start = std::time::Instant::now();
        let timeout = Duration::from_secs(5);

        loop {
            let status = self.get_status()?;
            if status[0] != STATUS_OK || status[4] == STATE_DFU_ERROR {
                return Err(format!(
                    "Write block {block_num} failed (status={})",
                    format_status(status)
                ));
            }
            match status[4] {
                STATE_DFU_IDLE | STATE_DFU_DOWNLOAD_IDLE => break,
                STATE_DFU_DNLOAD_SYNC | STATE_DFU_DOWNLOAD_BUSY => {
                    if start.elapsed() > timeout {
                        return Err(format!(
                            "Timeout exceeded while writing block {block_num} (status={})",
                            format_status(status)
                        ));
                    }
                    thread::sleep(Duration::from_millis(Self::poll_timeout_ms(status).max(10)));
                }
                other => {
                    return Err(format!(
                        "Write block {block_num} failed (unexpected DFU state 0x{other:02x}, status={})",
                        format_status(status)
                    ));
                }
            }
        }

        Ok(())
    }

    pub fn read_block(&mut self, block_num: u16, out: &mut [u8]) -> Result<usize, String> {
        self.control_in(DFU_UPLOAD, block_num, out, 500)
    }

    pub fn flash(
        &mut self,
        firmware: &[u8],
        address: u32,
        mut on_progress: impl FnMut(String),
    ) -> Result<(), String> {
        let _ = self.clear_status();
        on_progress("Starting mass erase...".to_string());
        self.mass_erase()?;
        on_progress("Mass erase complete. Setting address pointer...".to_string());
        self.set_address_pointer(address)?;
        on_progress("Address pointer set. Starting flash write...".to_string());

        let mut block_num: u16 = 2;
        let mut read_buffer = vec![0u8; BLOCK_SIZE];
        for chunk in firmware.chunks(BLOCK_SIZE) {
            on_progress(format!("Writing block {block_num}..."));
            self.write_block(block_num, chunk)?;

            on_progress(format!("Verifying block {block_num}..."));
            self.wait_upload_idle()?;

            let buf = &mut read_buffer[..chunk.len()];
            let n = self.read_block(block_num, buf)?;
            if n != chunk.len() {
                return Err(format!(
                    "Verification failed for block {block_num}: read {n} bytes, expected {}",
                    chunk.len()
                ));
            }
            if buf != chunk {
                return Err(format!("Error verifying block {}", block_num.saturating_sub(2)));
            }

            block_num = block_num.wrapping_add(1);
        }

        on_progress("Flash write completed successfully.".to_string());
        Ok(())
    }
}

fn format_status(status: [u8; 6]) -> String {
    let poll = DfuDevice::poll_timeout_ms(status);
    format!(
        "bStatus=0x{status:02x} bState=0x{state:02x} bwPollTimeout={poll} iString={istring}",
        status = status[0],
        state = status[4],
        poll = poll,
        istring = status[5]
    )
}

File: macos/EMWaver/EMWaver/Dfu.swift
import Foundation

// DFU implementation for STM32 ROM DFU (0483:DF11) backed by libusb (loaded dynamically).
// Mirrors crates/emwaver-dfu/src/lib.rs.

final class Dfu {
    static let vendorId: UInt16 = 0x0483
    static let productId: UInt16 = 0xDF11

    // Requests
    private static let DFU_DNLOAD: UInt8 = 0x01
    private static let DFU_UPLOAD: UInt8 = 0x02
    private static let DFU_GETSTATUS: UInt8 = 0x03
    private static let DFU_CLRSTATUS: UInt8 = 0x04
    private static let DFU_ABORT: UInt8 = 0x06

    // States
    private static let STATE_DFU_IDLE: UInt8 = 0x02
    private static let STATE_DFU_DNLOAD_SYNC: UInt8 = 0x03
    private static let STATE_DFU_DOWNLOAD_BUSY: UInt8 = 0x04
    private static let STATE_DFU_DOWNLOAD_IDLE: UInt8 = 0x05
    private static let STATE_DFU_MANIFEST_SYNC: UInt8 = 0x06
    private static let STATE_DFU_MANIFEST: UInt8 = 0x07
    private static let STATE_DFU_MANIFEST_WAIT_RESET: UInt8 = 0x08
    private static let STATE_DFU_UPLOAD_IDLE: UInt8 = 0x09
    private static let STATE_DFU_ERROR: UInt8 = 0x0A

    private static let STATUS_OK: UInt8 = 0x00

    static let blockSize: Int = 2048

    enum DfuError: Error, CustomStringConvertible {
        case libusb(String)
        case notFound
        case openFailed
        case protocolError(String)
        case timeout(String)

        var description: String {
            switch self {
            case .libusb(let s): return s
            case .notFound: return "No DFU device found (0483:DF11)."
            case .openFailed: return "Failed to open DFU device (0483:DF11)."
            case .protocolError(let s): return "DFU protocol error: \(s)"
            case .timeout(let s): return "Timeout: \(s)"
            }
        }
    }

    static func isConnected() -> Bool {
        do {
            let lib = try Libusb()
            var ctx: Libusb.libusb_context? = nil
            let rc = lib.libusb_init(&ctx)
            if rc != 0 { return false }
            defer { lib.libusb_exit(ctx) }
            let h = lib.libusb_open_device_with_vid_pid(ctx, vendorId, productId)
            if h != nil {
                lib.libusb_close(h)
                return true
            }
            return false
        } catch {
            return false
        }
    }

    static func openFirst() throws -> Dfu {
        let lib = try Libusb()
        var ctx: Libusb.libusb_context? = nil
        let rcInit = lib.libusb_init(&ctx)
        guard rcInit == 0 else { throw DfuError.libusb("libusb_init failed: \(rcInit)") }

        guard let handle = lib.libusb_open_device_with_vid_pid(ctx, Self.vendorId, Self.productId) else {
            lib.libusb_exit(ctx)
            throw DfuError.notFound
        }

        // Auto detach kernel driver where supported.
        _ = lib.libusb_set_auto_detach_kernel_driver(handle, 1)

        // STM32 ROM DFU typically uses interface 0. We can make this smarter later by scanning descriptors.
        let iface: Int32 = 0
        let rcClaim = lib.libusb_claim_interface(handle, iface)
        guard rcClaim == 0 else {
            lib.libusb_close(handle)
            lib.libusb_exit(ctx)
            throw DfuError.libusb("libusb_claim_interface(\(iface)) failed: \(rcClaim)")
        }

        let dfu = Dfu(lib: lib, ctx: ctx, handle: handle, interfaceNumber: UInt16(iface))

        // Best-effort clear error state like Rust.
        do {
            let st = try dfu.getStatus()
            if st.count >= 5, st[4] == STATE_DFU_ERROR {
                try? dfu.clearStatus()
            }
        } catch { }

        return dfu
    }

    private let lib: Libusb
    private var ctx: Libusb.libusb_context?
    private var handle: Libusb.libusb_device_handle?
    private let interfaceNumber: UInt16

    private init(lib: Libusb, ctx: Libusb.libusb_context?, handle: Libusb.libusb_device_handle, interfaceNumber: UInt16) {
        self.lib = lib
        self.ctx = ctx
        self.handle = handle
        self.interfaceNumber = interfaceNumber
    }

    func close() {
        if let h = handle {
            _ = lib.libusb_release_interface(h, Int32(interfaceNumber))
            lib.libusb_close(h)
        }
        handle = nil
        if let c = ctx {
            lib.libusb_exit(c)
        }
        ctx = nil
    }

    deinit { close() }

    // MARK: - DFU operations

    func flash(firmware: Data, address: UInt32, onProgress: @escaping (String, Double) -> Void) throws {
        try? clearStatus()

        let totalBlocks = Int(ceil(Double(firmware.count) / Double(Self.blockSize)))
        let totalSteps = max(1, totalBlocks * 2 + 2)
        var step: Int = 0
        func emit(_ msg: String) {
            let pct = min(100.0, (Double(step) * 100.0) / Double(totalSteps))
            onProgress(msg, pct)
        }

        emit("Starting mass erase...")
        try massErase()
        step += 1
        emit("Mass erase complete. Setting address pointer...")
        try setAddressPointer(address)
        step += 1
        emit("Address pointer set. Starting flash write...")

        let bytes = [UInt8](firmware)
        var blockNum: UInt16 = 2
        var readBuf = [UInt8](repeating: 0, count: Self.blockSize)

        for blockIndex in 0..<totalBlocks {
            let start = blockIndex * Self.blockSize
            let len = min(Self.blockSize, bytes.count - start)
            let chunk = Array(bytes[start..<(start + len)])

            emit("Writing block \(blockNum) (\(blockIndex + 1)/\(totalBlocks))...")
            try writeBlock(blockNum, chunk)
            step += 1

            emit("Verifying block \(blockNum) (\(blockIndex + 1)/\(totalBlocks))...")
            try waitUploadIdle()
            let n = try readBlock(blockNum, out: &readBuf, count: len)
            if n != len { throw DfuError.protocolError("Verification failed for block \(blockNum): read \(n) bytes, expected \(len)") }
            if !zip(readBuf.prefix(len), chunk).allSatisfy({ $0 == $1 }) {
                throw DfuError.protocolError("Error verifying block \(blockNum - 2)")
            }

            step += 1
            blockNum &+= 1
        }

        step = totalSteps
        emit("Flash write completed successfully.")
    }

    private func pollTimeoutMs(_ status: [UInt8]) -> UInt32 {
        if status.count < 4 { return 0 }
        return (UInt32(status[3]) << 16) | (UInt32(status[2]) << 8) | UInt32(status[1])
    }

    private func formatStatus(_ st: [UInt8]) -> String {
        if st.count < 6 { return "<short status len=\(st.count)>" }
        return String(format: "bStatus=0x%02X bState=0x%02X bwPollTimeout=%u iString=%u", st[0], st[4], pollTimeoutMs(st), st[5])
    }

    private func controlOut(_ request: UInt8, _ value: UInt16, _ data: [UInt8], timeoutMs: UInt32) throws {
        guard let h = handle else { throw DfuError.openFailed }

        // bmRequestType: 0x21 = Host->Device | Class | Interface
        var tmp = data
        let rc: Int32 = tmp.withUnsafeMutableBufferPointer { buf in
            let p = buf.baseAddress
            return lib.libusb_control_transfer(h, 0x21, request, value, interfaceNumber, p, UInt16(tmp.count), timeoutMs)
        }
        if rc < 0 {
            throw DfuError.libusb("libusb_control_transfer(OUT req=0x\(String(format: "%02X", request)) value=0x\(String(format: "%04X", value))) failed: \(rc)")
        }
    }

    private func controlIn(_ request: UInt8, _ value: UInt16, _ length: Int, timeoutMs: UInt32) throws -> [UInt8] {
        guard let h = handle else { throw DfuError.openFailed }

        var out = [UInt8](repeating: 0, count: length)
        let rc: Int32 = out.withUnsafeMutableBufferPointer { buf in
            let p = buf.baseAddress
            return lib.libusb_control_transfer(h, 0xA1, request, value, interfaceNumber, p, UInt16(length), timeoutMs)
        }
        if rc < 0 {
            throw DfuError.libusb("libusb_control_transfer(IN req=0x\(String(format: "%02X", request)) value=0x\(String(format: "%04X", value))) failed: \(rc)")
        }
        // rc is number of bytes transferred
        let n = Int(rc)
        if n < out.count { out.removeSubrange(n..<out.count) }
        return out
    }

    private func getStatus() throws -> [UInt8] {
        try controlIn(Self.DFU_GETSTATUS, 0, 6, timeoutMs: 500)
    }

    private func clearStatus() throws {
        try controlOut(Self.DFU_CLRSTATUS, 0, [], timeoutMs: 500)
    }

    private func abort() throws {
        try controlOut(Self.DFU_ABORT, 0, [], timeoutMs: 500)
    }

    private func waitDownloadIdle(timeout: TimeInterval = 5.0) throws {
        let deadline = Date().addingTimeInterval(timeout)
        while true {
            let st = try getStatus()
            let state = st.count >= 5 ? st[4] : 0
            if state == Self.STATE_DFU_IDLE || state == Self.STATE_DFU_DOWNLOAD_IDLE { return }
            if Date() > deadline {
                throw DfuError.timeout("waiting for download idle (status=\(formatStatus(st)))")
            }
            if state == Self.STATE_DFU_ERROR { try? clearStatus() }
            Thread.sleep(forTimeInterval: 0.05)
        }
    }

    private func waitUploadIdle(timeout: TimeInterval = 5.0) throws {
        let deadline = Date().addingTimeInterval(timeout)
        while true {
            let st = try getStatus()
            let state = st.count >= 5 ? st[4] : 0
            if state == Self.STATE_DFU_IDLE || state == Self.STATE_DFU_UPLOAD_IDLE { return }
            if Date() > deadline {
                throw DfuError.timeout("waiting for upload idle (status=\(formatStatus(st)))")
            }
            if state == Self.STATE_DFU_ERROR { try? clearStatus() }
            Thread.sleep(forTimeInterval: 0.05)
        }
    }

    private func massErase() throws {
        do {
            try massEraseOnce()
        } catch {
            _ = try? abort()
            _ = try? clearStatus()
            _ = try? waitDownloadIdle(timeout: 5.0)
            try massEraseOnce()
        }
    }

    private func massEraseOnce() throws {
        try waitDownloadIdle(timeout: 5.0)
        try controlOut(Self.DFU_DNLOAD, 0, [0x41], timeoutMs: 50)

        let deadline = Date().addingTimeInterval(60.0)
        while true {
            let st = try getStatus()
            let bStatus = st[0]
            let state = st[4]

            if bStatus != Self.STATUS_OK || state == Self.STATE_DFU_ERROR {
                throw DfuError.protocolError("Mass erase failed (status=\(formatStatus(st)))")
            }

            switch state {
            case Self.STATE_DFU_IDLE, Self.STATE_DFU_DOWNLOAD_IDLE:
                return

            case Self.STATE_DFU_DNLOAD_SYNC, Self.STATE_DFU_DOWNLOAD_BUSY,
                 Self.STATE_DFU_MANIFEST_SYNC, Self.STATE_DFU_MANIFEST, Self.STATE_DFU_MANIFEST_WAIT_RESET:
                if Date() > deadline {
                    throw DfuError.timeout("waiting for mass erase (status=\(formatStatus(st)))")
                }
                let sleepMs = max(10, Int(pollTimeoutMs(st)))
                Thread.sleep(forTimeInterval: Double(sleepMs) / 1000.0)

            default:
                throw DfuError.protocolError(String(format: "Mass erase failed (unexpected DFU state 0x%02X, status=%@)", state, formatStatus(st)))
            }
        }
    }

    private func setAddressPointer(_ address: UInt32) throws {
        try waitDownloadIdle(timeout: 5.0)

        let buf: [UInt8] = [
            0x21,
            UInt8(address & 0xFF),
            UInt8((address >> 8) & 0xFF),
            UInt8((address >> 16) & 0xFF),
            UInt8((address >> 24) & 0xFF),
        ]

        try controlOut(Self.DFU_DNLOAD, 0, buf, timeoutMs: 50)

        let deadline = Date().addingTimeInterval(5.0)
        while true {
            let st = try getStatus()
            let bStatus = st[0]
            let state = st[4]

            if bStatus != Self.STATUS_OK || state == Self.STATE_DFU_ERROR {
                throw DfuError.protocolError("Set address pointer failed (status=\(formatStatus(st)))")
            }

            switch state {
            case Self.STATE_DFU_IDLE, Self.STATE_DFU_DOWNLOAD_IDLE:
                return
            case Self.STATE_DFU_DNLOAD_SYNC, Self.STATE_DFU_DOWNLOAD_BUSY:
                if Date() > deadline {
                    throw DfuError.timeout("setting address pointer (status=\(formatStatus(st)))")
                }
                let sleepMs = max(10, Int(pollTimeoutMs(st)))
                Thread.sleep(forTimeInterval: Double(sleepMs) / 1000.0)
            default:
                throw DfuError.protocolError(String(format: "Set address pointer failed (unexpected DFU state 0x%02X, status=%@)", state, formatStatus(st)))
            }
        }
    }

    private func writeBlock(_ blockNum: UInt16, _ data: [UInt8]) throws {
        try waitDownloadIdle(timeout: 5.0)
        try controlOut(Self.DFU_DNLOAD, blockNum, data, timeoutMs: 500)

        let deadline = Date().addingTimeInterval(5.0)
        while true {
            let st = try getStatus()
            let bStatus = st[0]
            let state = st[4]

            if bStatus != Self.STATUS_OK || state == Self.STATE_DFU_ERROR {
                throw DfuError.protocolError("Write block \(blockNum) failed (status=\(formatStatus(st)))")
            }

            switch state {
            case Self.STATE_DFU_IDLE, Self.STATE_DFU_DOWNLOAD_IDLE:
                return
            case Self.STATE_DFU_DNLOAD_SYNC, Self.STATE_DFU_DOWNLOAD_BUSY:
                if Date() > deadline {
                    throw DfuError.timeout("writing block \(blockNum) (status=\(formatStatus(st)))")
                }
                let sleepMs = max(10, Int(pollTimeoutMs(st)))
                Thread.sleep(forTimeInterval: Double(sleepMs) / 1000.0)
            default:
                throw DfuError.protocolError(String(format: "Write block failed (unexpected DFU state 0x%02X, status=%@)", state, formatStatus(st)))
            }
        }
    }

    private func readBlock(_ blockNum: UInt16, out: inout [UInt8], count: Int) throws -> Int {
        try waitUploadIdle(timeout: 5.0)
        let data = try controlIn(Self.DFU_UPLOAD, blockNum, count, timeoutMs: 500)
        if out.count < count { out = [UInt8](repeating: 0, count: count) }
        for i in 0..<min(count, data.count) { out[i] = data[i] }
        return data.count
    }
}
